util/container/Dockerfile

```diff
@@ -4,174 +4,127 @@
+    tar xzf doxygen-${DOXYGEN_VERSION}.linux.bin.tar.gz && \
+    mv doxygen-${DOXYGEN_VERSION} doxygen
 
-# 2. Stage
-FROM ubuntu:22.04 AS snitch_cluster
-ARG SNITCH_LLVM_VERSION=latest
-ARG VERIBLE_VERSION=0.0-3318-g8d254167
+# Get the precompiled LLVM toolchain
+RUN latest_tag=`curl -s -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/pulp-platform/llvm-project/releases/latest | grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/'` && \
+echo "SNITCH_LLVM_VERSION=${SNITCH_LLVM_VERSION} LLVM_TAR=${LLVM_TAR} latest_tag=${latest_tag}" && \
+test "${SNITCH_LLVM_VERSION}" = "latest" && SNITCH_LLVM_VERSION=${latest_tag} || : ; \
+LLVM_TAR=riscv32-pulp-llvm-ubuntu2004-$(echo $SNITCH_LLVM_VERSION | cut -d '-' -f3-).tar.gz && \
+mkdir -p riscv-llvm && \
+echo "SNITCH_LLVM_VERSION=${SNITCH_LLVM_VERSION} LLVM_TAR=${LLVM_TAR} latest_tag=${latest_tag}" && \
+wget -qO- https://github.com/pulp-platform/llvm-project/releases/download/${SNITCH_LLVM_VERSION}/${LLVM_TAR} | \
+tar xvz --strip-components=1 -C riscv-llvm
+
+# 2. Stage: Filter python files to improve caching of the final container
+FROM ubuntu:${UBUNTU_VERSION} AS python_filter
+
+# Copy all files in `sw` and `util` and remove all non-python files
+COPY sw ./sw
+COPY util ./util
+
+RUN find sw -type f ! -name "*.py" -delete && \
+    find util -type f ! -name "*.py" -delete
+
+# 3. Stage: Create the final container
+FROM verilator/verilator:v${VERILATOR_VERSION} AS snitch_cluster
+ARG PYTHON_VERSION=3.11
+# Run dpkg without interactive dialogue
+ARG DEBIAN_FRONTEND=noninteractive
 
 LABEL version="0.1"
 LABEL description="Snitch container for hardware and software development."
 LABEL maintainer="zarubaf@iis.ee.ethz.ch"
-LABEL org.opencontainers.image.source https://github.com/pulp-platform/snitch_cluster
+LABEL org.opencontainers.image.source=https://github.com/pulp-platform/snitch_cluster
 
-WORKDIR /tools
+RUN apt update && \
+    apt install -y --no-install-recommends \
+        # General requirements
+        curl wget build-essential git \
+        # Required for python
+        software-properties-common gnupg2 \
+        # Required for fesvr
+        device-tree-compiler
 
-# Install (and cleanup) required packages (from apt-requirements.txt)
-# The list of extra packages is leftover from before this Dockerfile used
-# apt-requirements.txt
-#
-COPY apt-requirements.txt /tmp/apt-requirements.txt
-RUN apt-get update && \
-    sed 's/#.*//' /tmp/apt-requirements.txt \
-        | xargs apt-get install -y && \
-    apt-get install -y --no-install-recommends \
-        ca-certificates \
-        gnupg2 \
-        curl \
-        wget \
-        build-essential \
-        git && \
-    apt-get clean ; \
+# Install Python
+RUN apt-add-repository ppa:deadsnakes/ppa && \
+    apt update && \
+    apt install -y --no-install-recommends \
+        python${PYTHON_VERSION}-venv && \
+    apt clean && \
```

Comment:
We can potentially merge these two `RUN` commands right?

Reply:
Yes, they could be merged I believe